Sunburst panel

nickvasilyev opened the pull request:
I have a new sunburst panel. It looks similar to http://bl.ocks.org/kerryrodden/7090426 and is powered by solr facet pivot. Configuration involves entering a list of fields to run facet pivot on, like field1,field2,field3 and configuring a facet limit. In the future, I can implement a per field facet limit. 

When clicked on any slice of the doughnut, it will create filters based on the item and it's parents, then reload the dashboard. 

I did some testing on it, but couldn't spend as much time on it as I wanted to. A second pair of eyes would help.

Let me know what you think.  

![image 1](https://cloud.githubusercontent.com/assets/8037070/7169111/8dc5a8a4-e390-11e4-8d39-0136f8746f34.png)

![image](https://cloud.githubusercontent.com/assets/8037070/7169112/933b3bfa-e390-11e4-8fd7-0609305f250e.png)

nickvasilyev committed:
Added a sunburst panel

nickvasilyev committed:
changed field length

dougmcclure commented:
beautiful!  I'll give this a try.  I would love to learn how to create these!

dougmcclure commented:
I'm seeing nothing render in the panel.  I can see results from the query for all the facets. Any ideas?  Chrome is my browser, I cleared all cache after uploading your panel and making the change to config.js.

Tks!

Doug

snip
facet_counts": {
        "facet_queries": {},
        "facet_fields": {},
        "facet_dates": {},
        "facet_ranges": {},
        "facet_pivot": {
            "fieldA,fieldB": [
                [
                    "field",
                    "fieldA",
                    "value",
                    "BASLIGHT",
                    "count",
                    3245,
                    "pivot",
                    [
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Very Low",
                            "count",
                            1896
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Very High",
                            "count",
                            522
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Low",
                            "count",
                            492
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Medium",
                            "count",
                            334
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Medium High",
                            "count",
                            1
                        ]

nickvasilyev commented:
Great, thanks for testing this. 

Can you send me the full json output from Solr? You can find if you have chrome Dev tools open under the network section.

nickvasilyev commented:
Hi Doug, 

I fixed your data to look like facet.pivot output, ie:
var testdata =  {
"fieldA,fieldB": [
{
"field":"fieldA",
"value":"BASLIGHT",
"count":3245,
"pivot":[
{
"field":"fieldB",
"value":"Very Low",
"count":1896
},
{
"field":"fieldB",
"value":"Very High",
"count":522
},
{
"field":"fieldB",
"value":"Low",
"count":492
},
{
"field":"fieldB",
"value":"Medium",
"count":334},
{
"field":"fieldB",
"value":"Medium High",
"count":1
}]}]}

and was able to load the panel without a problem

![image](https://cloud.githubusercontent.com/assets/8037070/7207064/b61c7010-e504-11e4-8520-414c66e05c55.png)

dougmcclure commented:
how do we troubleshoot?

The data is there in the browser dev tools console query response.

nickvasilyev commented:
The data you posted does not look like facet.pivot json response. It has commas instead of semicolons and some nodes are lists instead of objects. Can you tell me where you got that from?

dougmcclure commented:
Here's a more complete one I hope (still sanitized). This is from the Chrome devtools console.

This system is using SolrCloud and a collection alias which represents all the collections in my environment.

https://10.0.0.1:9987/doug/solr/my_collection_alias/select?q=_datasource%3A%22Data%20Test%20Stream%22&fq=timestamp:[NOW/DAY-30DAY%20TO%20NOW/DAY%2B1DAY]&wt=json&facet=true&facet.pivot=fieldA,fieldB&facet.limit=1000

{
    "responseHeader": {
        "status": 0,
        "QTime": 1932,
        "params": {
            "facet": "true",
            "q": "_datasource:\"Data Test Stream\"",
            "facet.limit": "1000",
            "wt": "json",
            "facet.pivot": "fieldA,fieldB",
            "fq": "timestamp:[NOW/DAY-30DAY TO NOW/DAY+1DAY]"
        }
    },
    "response": {
        "numFound": 38912,
        "start": 0,
        "maxScore": 6.444475,
        "docs": [
            --snip--
        ]
    },
    "facet_counts": {
        "facet_queries": {},
        "facet_fields": {},
        "facet_dates": {},
        "facet_ranges": {},
        "facet_pivot": {
            "fieldA,fieldB": [
                [
                    "field",
                    "fieldA",
                    "value",
                    "BASLIGHT",
                    "count",
                    38912,
                    "pivot",
                    [
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Very Low",
                            "count",
                            13467
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Low",
                            "count",
                            11778
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Medium",
                            "count",
                            8909
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Very High",
                            "count",
                            3623
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Medium High",
                            "count",
                            847
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "High",
                            "count",
                            288
                        ]
                    ]
                ]
            ]
        }
    }
}

dougmcclure commented:
sorry, it's also Solr 4.7.1.

nickvasilyev commented:
There is something wrong with your Solr Output. 

This is the output I am getting on solr cloud 4.6 with a collection alias, with some notes on why your output is different. 

 "facet_pivot":{
      "fieldA,fieldB":[{  <---- See how there is a squiggly bracket here, you have a square one instead
          "field":"fieldA", <--- See how for each one of these, there is a colon, not a comma separating 
          "value":"00000000",      ---- keys and values
          "count":254628,
          "pivot":[{           <--- here again, square bracket followed by squiggly 
              "field":"FieldB",
              "value":"VALA",
              "count":2},
            {
              "field":"FieldB",
              "value":"VALB",
              "count":2}]},

dougmcclure commented:
Any ideas on what would cause that? A change in 4.7.1 maybe?

nickvasilyev commented:
Hi Doug, 
No idea, can you visually confirm this by querying solr directly? It is very weird,  because the output you posted doesn't conform to json standards, so everything would break if Solr was outputting it. Do other panels that rely on facet pivot like heatmap work for you?

dougmcclure commented:
I've never gotten heatmap to work. I run the query string in browser to
solr direct and the results look same to my eyes:

{
    "responseHeader": {
        "status": 0,
        "QTime": 53,
        "params": {
            "facet": "true",
            "q": "_datasource:\"blah\"",
            "facet.limit": "1000",
            "wt": "json",
            "facet.pivot": "fieldA,fieldB",
            "fq": "timestamp:[NOW/DAY-30DAY TO NOW/DAY+1DAY]"
        }
    },
    "response": {
        "numFound": 38912,
        "start": 0,
        "maxScore": 6.444475,
        "docs": [

snip
        ]
    },
    "facet_counts": {
        "facet_queries": {},
        "facet_fields": {},
        "facet_dates": {},
        "facet_ranges": {},
        "facet_pivot": {
            "fieldA,fieldB": [
                [
                    "field",
                    "fieldA",
                    "value",
                    "BASLIGHT",
                    "count",
                    38912,
                    "pivot",
                    [
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Very Low",
                            "count",
                            13467
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Low",
                            "count",
                            11778
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Medium",
                            "count",
                            8909
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Very High",
                            "count",
                            3623
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "Medium High",
                            "count",
                            847
                        ],
                        [
                            "field",
                            "fieldB",
                            "value",
                            "High",
                            "count",
                            288
                        ]
                    ]
                ]
            ]
        }
    }
}

On Fri, Apr 17, 2015 at 2:59 PM, Nick Vasilyev notifications@github.com
wrote:

> Hi Doug,
> No idea, can you visually confirm this by querying solr directly? It is
> very weird, because the output you posted doesn't conform to json
> standards, so everything would break if Solr was outputting it. Do other
> panels that rely on facet pivot like heatmap work for you?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/LucidWorks/banana/pull/169#issuecomment-94052963.

aadel commented:
@wrdrvr Very nice! Got this running without any problems on 4.9.0. I think it would be nice to have an option for legend and show labels on segments. There's a small glitch that keeps tool-tips permanently visible after adding filters (clicking on segments).

@dougmcclure the output above is weird, not JSON format. Could be something specific to pivot and aliasing in 4.7.1?

nickvasilyev commented:
Hi Ahmed, 
Thanks, I will into the glitch. As for labels and legend, I think it would be nice too, but I am not sure where to place it, especially considering the large number of possible values.

dougmcclure commented:
thanks- I'll ask our dev guys if we've done anything to a stock Solr 4.7.1
package in the faceting/pivoting area.

Nick - if you have any more works of art like this, I'd love to help
somehow.  I've been very keen on the d3 sankey diagrams and think they'd be
very cool for visualizing data moving from source to destinations, user
transaction flows, etc.

On Fri, Apr 17, 2015 at 4:13 PM, Nick Vasilyev notifications@github.com
wrote:

> Hi Ahmed,
> Thanks, I will into the glitch. As for labels and legend, I think it would
> be nice too, but I am not sure where to place it, especially considering
> the large number of possible values.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/LucidWorks/banana/pull/169#issuecomment-94065274.

nickvasilyev committed:
Fixed persistent tooltip, set rows to 0 in query

Fixed tooltips that were hanging around after reload and added rows=0 to
Solr query.

nickvasilyev committed:
Removed unused code and fixed the loading indicator

Removed a no longer needed function and fixed loading indicator.

nickvasilyev commented:
@aadel I was able to replicate the persistent tooltip by having one display while the panel reloaded. I am not sure why it was hanging around, but I am removing all sunburst tooltips on re-render now. I think this should fix it. 

Also, I noticed a problem with the loading indicator. It was only on while the data was being fetched, but it may take a few seconds to render the panel, so I moved it in the D3 code. 

@dougmcclure 
That is a cool chart, I would like to see something like this as well. I will try to play with it a bit once I have some free time.

aadel commented:
@wrdrvr I think legend can go on any of the four directions (similar to terms panel) but as you mentioned values will be most probably too many especially when there are several layers (circles), the legend is just nice to have and can be hidden by default.

nickvasilyev commented:
@aadel Ok, I will take a stab at it when I have a bit more time.

aadel commented:
@wrdrvr confirmed the tooltip issue is fixed. Thanks for the contribution!

@dougmcclure +1 for having Sankey diagrams. They are very useful for analyzing flows such as transactions or energy and would be cool have.
